File: src/main/scala/ServletRequest2Stream.scala
import java.util.concurrent.atomic.AtomicReference
import javax.servlet.{ServletInputStream, ReadListener, ServletRequest}
import org.reactivestreams.api.{Consumer, Producer}
import org.reactivestreams.spi.{Subscription, Subscriber, Publisher}

object ServletRequest2Stream {

  def apply(sr: ServletRequest, callback: Producer[Byte] => Unit) = new ServletRequest2Stream(sr, callback)

  class MySubscription(val subscriber: Subscriber[Byte],
                       input: ServletInputStream,
                       cancellation: () => Unit) extends Subscription {

    subscriber.onSubscribe(this)

    def readAndSend(elements: Int): Unit = {
      val buffer = Array.ofDim[Byte](elements)
      val readElements = input.read(buffer)
      if (readElements > 0) {
        buffer.take(readElements).foreach(subscriber.onNext)
      } else {
        subscriber.onComplete()
        cancel()
      }
    }

    override def requestMore(elements: Int): Unit = {
      if (input.isReady) {
        readAndSend(elements)
      }
    }

    override def cancel(): Unit = {
      cancellation()
    }

  }

}

class ServletRequest2Stream(sr: ServletRequest, callback: Producer[Byte] => Unit) {

  import ServletRequest2Stream.MySubscription

  private val input = sr.getInputStream
  private val currentSubscription: AtomicReference[Option[MySubscription]] = new AtomicReference[Option[MySubscription]]()
  currentSubscription.set(None)
  private val readListener = new ReadListener {

    override def onDataAvailable(): Unit = {
      currentSubscription.get match {
        case Some(subscription) =>
          subscription.readAndSend(1)
        case None =>
      }
    }

    override def onAllDataRead(): Unit = {
      currentSubscription.get match {
        case Some(subscription) =>
          subscription.subscriber.onComplete()
          currentSubscription.set(None)
        case None =>
      }

    }

    override def onError(t: Throwable): Unit = {
      currentSubscription.get match {
        case Some(subscription) =>
          subscription.subscriber.onError(t)
        case None =>
      }

    }

  }

  input.setReadListener(readListener)

  private val producer = new Producer[Byte] {

    override def getPublisher: Publisher[Byte] = {
      new Publisher[Byte] {

        override def subscribe(subscriber: Subscriber[Byte]): Unit = {
          currentSubscription.set(Some(new MySubscription(subscriber, input, () => currentSubscription.set(None) )))
          readListener.onDataAvailable()
        }
      }
    }

    override def produceTo(consumer: Consumer[Byte]): Unit = {
      getPublisher.subscribe(consumer.getSubscriber)
    }

  }

  callback(producer)

}

File: src/main/java/ServletRequest2Stream.java
import org.reactivestreams.api.Consumer;
import org.reactivestreams.api.Producer;
import org.reactivestreams.spi.Publisher;
import org.reactivestreams.spi.Subscriber;
import org.reactivestreams.spi.Subscription;

import javax.servlet.ReadListener;
import javax.servlet.ServletInputStream;
import javax.servlet.ServletRequest;
import java.io.IOException;
import java.util.concurrent.atomic.AtomicReference;

public final class ServletRequest2Stream {

    private static final class MySubscription implements Subscription {

        private final Subscriber<Byte> subscriber;
        private final ServletInputStream input;
        private final Runnable cancellation;

        public MySubscription(final Subscriber<Byte> subscriber,
                              final ServletInputStream input,
                              final Runnable cancellation) {
            this.subscriber = subscriber;
            this.input = input;
            this.cancellation = cancellation;
            subscriber.onSubscribe(this);
        }

        public void readAndSend(final int elements) {
            byte[] buffer = new byte[elements];
            int readElements;
            try {
                readElements = input.read(buffer);
            } catch (IOException e) {
                subscriber.onError(e);
                return;
            }
            if (readElements > 0) {
                for (int i = 0; i < readElements; i++) {
                    subscriber.onNext(buffer[i]);
                }
            } else {
                subscriber.onComplete();
                cancel();
            }
        }

        @Override
        public void requestMore(final int elements) {
            if (input.isReady()) {
                readAndSend(elements);
            }
        }

        @Override
        public void cancel() {
            cancellation.run();
        }

    }

    private final ServletInputStream input;
    private final AtomicReference<MySubscription> currentSubscription = new AtomicReference<>(null);
    private final ReadListener readListener = new ReadListener() {

        @Override
        public void onDataAvailable() throws IOException {
            MySubscription current = currentSubscription.get();
            if (current != null) {
                current.readAndSend(1);
            }
        }

        @Override
        public void onAllDataRead() throws IOException {
            MySubscription current = currentSubscription.get();
            if (current != null) {
                current.subscriber.onComplete();
            }

        }

        @Override
        public void onError(final Throwable t) {
            MySubscription current = currentSubscription.get();
            if (current != null) {
                current.subscriber.onError(t);
            }
        }
    };

    private final Producer<Byte> producer = new Producer<Byte>() {

        @Override
        public Publisher<Byte> getPublisher() {
            return new Publisher<Byte>() {
                public void subscribe(final Subscriber<Byte> subscriber) {
                    Runnable cancellation = new Runnable() {
                        public void run() {
                            currentSubscription.set(null);
                        }
                    };
                    currentSubscription.set(new MySubscription(subscriber, input, cancellation));
                    try {
                        readListener.onDataAvailable();
                    } catch (IOException ignored) {}
                }
            };
        }

        @Override
        public void produceTo(final Consumer<Byte> consumer) {
            getPublisher().subscribe(consumer.getSubscriber());
        }
    };

    public ServletRequest2Stream(final ServletRequest sr) throws IOException {
        input = sr.getInputStream();
        input.setReadListener(readListener);
    }

    public Producer<Byte> getProducer() {
        return producer;
    }

}
